### DISCUSSION TITLE
tutorials/intro-typescript/14-mutations

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#121] [course feedback]
# tutorials/intro-typescript/14-mutations

https://www.apollographql.com/tutorials/intro-typescript/14-mutations

**Comment by jjbindr** [comment]
I think the import to spotify-api.ts at this point is 

``import { PlaylistModel, TrackModel, SnapshotOrError } from '../models'``

**Reply by AlexRebula** [reply]
Indeed. They have not updated that in the tutorial.

**Comment by nlevchuk** [comment]
In the "Handling the sad path" chapter, I was able to create a listing with numOfBeds = 0. However, I got the "sad" response when I updated the attribute to `-1`

**Reply by mabuyo** [reply]
Ah, good catch @nlevchuk - we missed something in our data source logic. Fixing shortly, thanks for flagging!

**Reply by nlevchuk** [reply]
@mabuyo You are welcome!:)

**Comment by stevegoossens** [comment]
Just a minor wording/"copy" issue:

> Something went wrong: 503: Service Unavailable

This is the note about the _500_ response added for the "sad" response. It should probably be altered to:

> Something went wrong: 500: Internal Server Error

**Comment by cunisia** [comment]
Really nice course!

**Comment by DaleSeo** [comment]
I suggest the following change for the `Building the ListingAPI Method` section:

```diff
createListing(listing: CreateListingInput): Promise<Listing> {
-  return this.post("listings", {
+  return this.post<Listing>("listings", {
    body: {
      listing
    }
  });
}
```

**Comment by DaleSeo** [comment]
I found a little typo in the `Handling the sad path` section:

```diff
-When we run the operation, we'll see the values we expect: the mutation failed, and our error message comes through loud and clear. `Something went wrong: 503: Service Unavailable`!
+When we run the operation, we'll see the values we expect: the mutation failed, and our error message comes through loud and clear. `Something went wrong: Please provide a valid number of beds`!
```

**Reply by lizhennessy** [reply]
Thanks for the catch @DaleSeo ! Fix is on the way.

**Reply by DaleSeo** [reply]
Thanks for addressing my feedback, @lizhennessy!

**Comment by otlyn** [comment]
Hello, could someone explain why we use `Promise<Listing>` instead of `Promise<CreateListingResponse>`?

```
createListing(listing: CreateListingInput): Promise<Listing> {
  return this.post("listings", {
    body: {
      listing
    }
  });
}
```

**Reply by lizhennessy** [reply]
Hi there @otlyn! This method actually just reflects what we get back from the REST API, rather than what we return from our resolvers. We've encoded our schema to say that the mutation takes in a `Listing` object and returns a `CreateListingResponse` object; it's the job of our `Mutation.createListing` resolver to actually return an object that satisfies the `CreateListingResponse` shape!

If we run this mutation in Explorer (and log out the response from calling our `ListingAPI` method `createListing`) we can see the response from the REST API. It's actually just the newly created listing! This gets returned by our `createListing` method here, and ends up in our resolver function. Thereafter we write our resolver logic to return an object that takes the shape of `CreateListingResponse`.

**Reply by otlyn** [reply]
Great, thank you for clarifying.

**Comment by otlyn** [comment]
There's a very small formatting issue under the 'show code for `Mutation.createListing`'. 
Here's the fix: 
```
Mutation: {
    createListing: async (_, { input }, { dataSources }) => {
      try {
        const response = await dataSources.listingAPI.createListing(input);
        return {
          code: 200,
          success: true,
          message: "Listing successfully created!",
          listing: response,
        };
      } catch (err) {
        return {
          code: 500,
          success: false,
          message: `Something went wrong: ${err.extensions.response.body}`,
          listing: null,
        };
      }
    },
  },
```

**Reply by lizhennessy** [reply]
Thanks for pointing this out @otlyn! Fix is on the way.

**Comment by ddube95** [comment]
Hello,
The solution to the last quiz is given as:

type Mutation {
  createPlanet(input: CreatePlanetInput!): CreatePlanetResponse!
}

type CreatePlanetInput {
  "The planet's name"
  name: String!
  "The total estimated mass of the planet (in kg)"
  mass: Int
}
........

Shouldn't 
 type CreatePanetInput { 
...
 } 

rather be  

input CreatePanetInput {
.... 
}

**Reply by mabuyo** [reply]
Good catch @ddube95 - we should be using an `input` type to follow best practice (though the schema is still valid!). We'll make this change shortly.

**Comment by samblackmore** [comment]
Looks like this copy isn't accurate:
> To test our sad path, let's try creating a listing with a non-existent amenity ID.

Instead we test the sad path by setting `numOfBeds` to zero

Great course btw!
